src/sysfiles.c

```diff
@@ -3598,6 +3598,56 @@ Obj SyReadStringFid(Int fid) {
 #endif
 
 
+#ifdef SYS_IS_64_BIT
+void * SyMemmove(void * dst, const void * src, UInt size)
+{
+    return memmove(dst, src, size);
+}
+#else
+// The memmove in glibc on 32-bit SSE2 systems, contained in
+// __memmove_sse2_unaligned, is buggy in at least versions
+// 2.21 - 2.25 when crossing the 2GB boundary, so GAP must
+// include it's own simple memmove implementation.
+void * SyMemmove(void * dst, const void * src, UInt size)
+{
+    char *       d = dst;
+    const char * s = src;
+
+    if ((dst == src) || (size == 0))
+        return dst;
+
+    if (d < s || d > s + size) {
+        memcpy(dst, src, size);
+    }
+    else {
+        d = d + (size - 1);
+        s = s + (size - 1);
+
+        // This is about 4x slower than glibc, but
+        // is simple and also complicated enough that
+        // gcc or clang seem unable to "optimise" it back
+        // into a call to memmove.
+
+        // Do size 4 jumps
+        while (size > 4) {
+            *d = *s;
+            *(d - 1) = *(s - 1);
+            *(d - 2) = *(s - 2);
+            *(d - 3) = *(s - 3);
+            d -= 4;
+            s -= 4;
+            size -= 4;
+        }
+        // Finish
+        while (size > 0) {
+            *d-- = *s--;
+            size--;
+        }
```

Comment:
I don't understand how this code can be correct: normally, any `memmove` implementation needs two code paths: one which moves data "forward", another which moves its "backwards". One deals with `memmove(ptr+1, ptr, size)` one with `memmove(ptr-1, ptr, size)`.

Looking at it, I think this code here always copies backwards, which would mean `memmove(ptr-1, ptr, size)` produces garbage. Am I missing something?

Reply:
Once again, we could really benefit from test setup for pure C code, so that one can write unit tests for functions like this.